#
# Provide optional sftp idle session timeout for ZFSSA only.  
# Not documented for Solaris.
#
# The sftp-server changes are rejected by upstream in favor of a longer-term,
# more-comprehensive idle session timeout that is in development.  So the
# upstream fix, when it is published, will supplant this temporary fix
# as soon as is feasible within the ZFSSA release cycle.
#
# The interface to sftp MUST NOT BE CHANGED here without coordinating
# with ZFSSA.
#
# The changes to sftp.c are from a suggested fix of upstream bug bz2719,
# they are not visible to the user because of privilege separation,  but will
# become so if the separation is restructured to allow better messaging.
#
diff -rupN old/sftp-server.c new/sftp-server.c
--- old/sftp-server.c	2017-11-30 16:16:30.241034978 +0000
+++ new/sftp-server.c	2018-01-05 09:34:57.941986733 +0000
@@ -39,6 +39,7 @@
 #include <time.h>
 #include <unistd.h>
 #include <stdarg.h>
+#include <signal.h>
 
 #include "xmalloc.h"
 #include "sshbuf.h"
@@ -92,6 +93,9 @@ struct Stat {
 	Attrib attrib;
 };
 
+/* sftp idle timeout */
+static volatile sig_atomic_t g_timed_out = 0;
+
 /* Packet handlers */
 static void process_open(u_int32_t id);
 static void process_close(u_int32_t id);
@@ -1852,7 +1856,10 @@ sftp_server_cleanup_exit(int i)
 		    pw->pw_name, client_addr);
 	}
 #ifdef SFTP_AUDIT
-	audit_sftp_session_stop((i != 0) ? ADT_FAIL_VALUE_PROGRAM : 0);
+	if (i == 3)
+	    audit_sftp_session_stop(ADT_FAIL_VALUE_IDLE_TIMEOUT);
+	else
+	    audit_sftp_session_stop((i != 0) ? ADT_FAIL_VALUE_PROGRAM : 0);
 #endif
 	_exit(i);
 }
@@ -1871,11 +1878,18 @@ sftp_server_usage(void)
 	exit(1);
 }
 
+static void
+sftp_alarm_handler(int sig)
+{
+	g_timed_out = 1;
+}
+
 int
 sftp_server_main(int argc, char **argv, struct passwd *user_pw)
 {
 	fd_set *rset, *wset;
 	int i, r, in, out, max, ch, skipargs = 0, log_stderr = 0;
+	int timeout = 0;
 	ssize_t len, olen, set_size;
 	SyslogFacility log_facility = SYSLOG_FACILITY_AUTH;
 	char *cp, *homedir = NULL, buf[4*4096];
@@ -1891,7 +1905,7 @@ sftp_server_main(int argc, char **argv,
 	pw = pwcopy(user_pw);
 
 	while (!skipargs && (ch = getopt(argc, argv,
-	    "d:f:l:P:p:Q:u:cehR")) != -1) {
+	    "d:f:l:P:p:Q:t:u:cehR")) != -1) {
 		switch (ch) {
 		case 'Q':
 			if (strcasecmp(optarg, "requests") != 0) {
@@ -1951,6 +1965,20 @@ sftp_server_main(int argc, char **argv,
 				fatal("Invalid umask \"%s\"", optarg);
 			(void)umask((mode_t)mask);
 			break;
+		case 't':
+			timeout = atoi(optarg);
+			debug("SFTP: optarg = %s\n", optarg);
+			debug("SFTP: setting timeout to %d\n", timeout);
+			if (timeout < 0) {
+				/*
+				 * -t0 is redundant, but will be used
+				 * by ZFSSA BUI/CLI
+				 */
+				error("Invalid timeout: \"%s\". Ignored.",
+				    optarg);
+				timeout = 0;
+			}
+			break;
 		case 'h':
 		default:
 			sftp_server_usage();
@@ -2022,6 +2050,7 @@ sftp_server_main(int argc, char **argv,
 	}
 
 	set_size = howmany(max + 1, NFDBITS) * sizeof(fd_mask);
+	signal(SIGALRM, sftp_alarm_handler);
 	for (;;) {
 		memset(rset, 0, set_size);
 		memset(wset, 0, set_size);
@@ -2043,13 +2072,23 @@ sftp_server_main(int argc, char **argv,
 		if (olen > 0)
 			FD_SET(out, wset);
 
+		if (timeout > 0)
+			alarm(timeout);
+
 		if (select(max+1, rset, wset, NULL, NULL) < 0) {
+			if (g_timed_out){
+				debug("SFTP idle timeout, closing session.");
+				sftp_server_cleanup_exit(3);
+			}
 			if (errno == EINTR)
 				continue;
 			error("select: %s", strerror(errno));
 			sftp_server_cleanup_exit(2);
 		}
 
+		if (timeout > 0)
+			alarm(0);
+
 		/* copy stdin to iqueue */
 		if (FD_ISSET(in, rset)) {
 			len = read(in, buf, sizeof buf);
diff -rupN old/sftp.c new/sftp.c
--- old/sftp.c	2017-12-01 15:29:31.649735226 +0000
+++ new/sftp.c	2017-12-01 15:29:39.739069916 +0000
@@ -253,6 +253,24 @@ cmd_interrupt(int signo)
 	errno = olderrno;
 }
 
+/*ARGSUSED*/
+static void
+sigchld_handler(int sig)
+{
+	int save_errno = errno;
+	pid_t pid;
+	const char msg[] = "\rConnection closed.  \n";
+
+	/* Report if ssh transport process dies. */
+	while ((pid = waitpid(sshpid, NULL, WNOHANG|WNOWAIT)) == -1 &&
+	    errno == EINTR)
+		continue;
+	if (pid == sshpid)
+		(void)write(STDERR_FILENO, msg, sizeof(msg) - 1);
+
+	errno = save_errno;
+}
+
 static void
 help(void)
 {
@@ -2227,6 +2227,7 @@ interactive_loop(struct sftp_conn *conn,
 		if (err != 0)
 			break;
 	}
+	signal(SIGCHLD, SIG_DFL);
 	free(remote_path);
 	free(startdir);
 	free(conn);
@@ -2293,6 +2312,7 @@ connect_to_server(char *path, char **arg
 	signal(SIGTERM, killchild);
 	signal(SIGINT, killchild);
 	signal(SIGHUP, killchild);
+	signal(SIGCHLD, sigchld_handler);
 	signal(SIGTSTP, suspchild);
 	signal(SIGTTIN, suspchild);
 	signal(SIGTTOU, suspchild);
